drivers/adc/adc_lpc_adc.c

```diff
@@ -0,0 +1,259 @@
+{
+	struct lpc_adc_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	error = lpc_adc_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+static int lpc_adc_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	return lpc_adc_read_async(dev, sequence, NULL);
+}
+
+static void lpc_adc_start_channel(const struct device *dev)
+{
+	const struct lpc_adc_config *config = dev->config;
+	struct lpc_adc_data *data = dev->data;
+	adc_conv_seq_config_t adc_conv_seq;
+	ADC_Type *base = config->base;
+
+	data->channel_id = find_lsb_set(data->channels) - 1;
+
+	adc_conv_seq.channelMask = (1U << data->channel_id);
+	adc_conv_seq.triggerMask = 0U;
+	adc_conv_seq.triggerPolarity = kADC_TriggerPolarityPositiveEdge;
+	adc_conv_seq.enableSingleStep = false;
+	adc_conv_seq.enableSyncBypass = false;
+	adc_conv_seq.interruptMode = kADC_InterruptForEachSequence;
+	ADC_SetConvSeqAConfig(base, &adc_conv_seq);
+	ADC_EnableConvSeqA(base, true);
+
+	ADC_DoSoftwareTriggerConvSeqA(base);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct lpc_adc_data *data =
+		CONTAINER_OF(ctx, struct lpc_adc_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	lpc_adc_start_channel(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+			  bool repeat_sampling)
+{
+	struct lpc_adc_data *data =
+		CONTAINER_OF(ctx, struct lpc_adc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void lpc_adc_isr(const struct device *dev)
+{
+	const struct lpc_adc_config *config = dev->config;
+	struct lpc_adc_data *data = dev->data;
+	ADC_Type *base = config->base;
+	adc_result_info_t adc_result;
+
+	data->channel_id = find_lsb_set(data->channels) - 1;
+	ADC_ClearStatusFlags(base, kADC_ConvSeqAInterruptFlag);
+	ADC_GetChannelConversionResult(base, data->channel_id, &adc_result);
+
+	*data->buffer++ = adc_result.result;
+	data->channels &= ~BIT(data->channel_id);
+
+	if (data->channels) {
+		lpc_adc_start_channel(dev);
+	} else {
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static const struct adc_driver_api lpc_adc_driver_api = {
+	.channel_setup = lpc_adc_channel_setup,
+	.read = lpc_adc_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = lpc_adc_read_async,
+#endif
+};
+
+static int lpc_adc_init(const struct device *dev)
+{
+	const struct lpc_adc_config *config = dev->config;
+	struct lpc_adc_data *data = dev->data;
+	ADC_Type *base = config->base;
+	adc_config_t adc_config;
+	uint32_t frequency = 0;
+
+	POWER_DisablePD(kPDRUNCFG_PD_ADC0);
+	POWER_DisablePD(kPDRUNCFG_PD_VD7_ENA);
+	POWER_DisablePD(kPDRUNCFG_PD_VREFP_SW);
+	POWER_DisablePD(kPDRUNCFG_PD_TEMPS);
+
+	/* Enable Clock for ADC0 channel */
+	CLOCK_EnableClock(kCLOCK_Adc0);
+
+	base->CTRL |= ADC_CTRL_BYPASSCAL_MASK;
+	frequency = CLOCK_GetFreq(kCLOCK_BusClk);
```

Comment:
This should be using the `clock_control` driver API. You would need to add device tree properties for the adc. Please refer the LPC UART driver.

Reply:
I didn't get your point. How to define clock frequency in dts file?

Reply:
Below is an example of how we get clock frequency for the I2C driver.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_mcux_flexcomm.c#L65

Below is an example of how we specify in the device tree which clock
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_lpc55S1x_common.dtsi#L150

Below is the clock control driver for LPC:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/clock_control/clock_control_mcux_syscon.c

Hope this helps.

Reply:
Thanks for the hint.

Reply:
@mmahadevan108 the current **clock_control** driver support's only to get the frequency for peripheral clocks like flexcomm, can, timers.
But for this adc driver, **BusClock** is used to get the clock frequency. 

For that we need to introduce a new macro in a header file **dt-bindings/clock/mcux_lpc_syscon_clock.h**  to get the clock frequency of system clock's.

Is this approach is correct?

Reply:
Yes, that would be the approach to take.